scio-test/src/main/scala/com/spotify/scio/testing/CoderAssertions.scala

```diff
@@ -30,102 +32,175 @@ import scala.reflect.ClassTag
-        beamCoder.consistentWithEquals() shouldBe true
-      }
+  def materializeToTransformOf[ExpectedBeamCoder: ClassTag]: CoderAssertion =
+    ctx => {
+      ctx.beamCoder shouldBe a[MaterializedCoder[_]]
+      ctx.beamCoder.asInstanceOf[MaterializedCoder[_]].bcoder shouldBe a[TransformCoder[_, _]]
+      val innerCoder =
+        ctx.beamCoder.asInstanceOf[MaterializedCoder[_]].bcoder.asInstanceOf[TransformCoder[_, _]]
+      innerCoder.bcoder shouldBe a[ExpectedBeamCoder]
     }
 
-  def beDeterministic[T: ClassTag](
-    opts: PipelineOptions = DefaultPipelineOptions
-  ): CoderAssertion[T] =
-    new CoderAssertion[T] {
-      override def assert(value: T)(implicit c: Coder[T], eq: Equality[T]): Assertion = {
-        val beamCoder = CoderMaterializer.beamWithDefault(c, o = opts)
-        noException should be thrownBy beamCoder.verifyDeterministic()
-      }
+  /*
+   * Checks that Beam's registerByteSizeObserver() and encode() are consistent
+   * */
+  def bytesCountTested[T <: Object: ClassTag](): CoderAssertionT[T] =
+    ctx => {
+      val arr = Array(ctx.actualValue.get)
+      noException should be thrownBy CoderProperties.testByteCount(
+        ctx.beamCoder,
+        BCoder.Context.OUTER,
+        arr
+      )
     }
 
-  def coderIsSerializable[A](implicit c: Coder[A]): Assertion =
-    coderIsSerializable(CoderMaterializer.beamWithDefault(c))
+  /**
+   * Verifies that for the given coder and values, the structural values are equal if and only if
+   * the encoded bytes are equal. Verifies for Outer and Nested contexts
+   */
+  def structuralValueConsistentWithEquals(): CoderAssertion = ctx => {
+    noException should be thrownBy CoderProperties.structuralValueConsistentWithEquals(
+      ctx.beamCoder,
+      ctx.actualValue.get,
+      ctx.actualValue.get
+    )
+  }
 
-  private def coderIsSerializable[A](beamCoder: BCoder[A]): Assertion =
-    noException should be thrownBy SerializableUtils.ensureSerializable(beamCoder)
+  /** Passes all checks on Beam coder */
+  def beFullyCompliant[T <: Object: ClassTag](): CoderAssertionT[T] = ctx => {
+    structuralValueConsistentWithEquals()(ctx)
+    beSerializable()(ctx)
+    beConsistentWithEquals()(ctx)
+    bytesCountTested[T]().apply(ctx)
+    beDeterministic()(ctx)
+  }
 
-  private def checkRoundtripWithCoder[T](beamCoder: BCoder[T], value: T)(implicit
-    eq: Equality[T]
-  ): Assertion = {
-    val bytes = CoderUtils.encodeToByteArray(beamCoder, value)
-    val result = CoderUtils.decodeFromByteArray(beamCoder, bytes)
+  def beFullyCompliantNonDeterministic[T <: Object: ClassTag](): CoderAssertionT[T] =
+    ctx => {
+      structuralValueConsistentWithEquals()(ctx)
+      beSerializable()(ctx)
+      beConsistentWithEquals()(ctx)
+      bytesCountTested[T]().apply(ctx)
+      beNonDeterministic()(ctx)
+    }
 
-    result should ===(value)
-  }
+  def beFullyCompliantNotConsistentWithEquals[T <: Object: ClassTag](): CoderAssertionT[T] =
```

Comment:
do we need `T <: Object` ?

Reply:
Unfortunately yes, it is needed by a nested check `def bytesCountTested[T <: Object: ClassTag]()` that requires it for:
```
val arr = Array(ctx.actualValue.get)
```
where `.get` returns `T`